src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/BufferInternal.cs

```diff
@@ -0,0 +1,271 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+
+namespace System {
+
+    // These sources are taken from corclr repo (src\mscorlib\src\System\Buffer.cs)
+    // The reason for this duplication is that System.Runtime.dll 4.0.10 did not expose Buffer.MemoryCopy,
+    // but we need to make this component work with System.Runtime.dll 4.0.10
+    // The methods AreOverlapping and SlowCopyBackwards are not from Buffer.cs. Buffer.cs does an internal CLR call for these.
+    static class BufferInternal
+    {       
+        // This method has different signature for x64 and other platforms and is done for performance reasons.
+        [System.Security.SecurityCritical]
+#if WIN64
```

Comment:
Have you tested that it'll do the right thing when compiled as AnyCPU w/o win64 defined and JIT'ed as 64-bit?  I did a quick scan and it seems that this would be a) slower than it has to be b) restricted to uint lengths.  I couldn't find any correctness issues but I didn't scan everything.

IIRC this piece of code was very finely tuned to do the right tradeoff between perf of native vs overhead of p-invoke.  Using 32bit always might invalidate that.  We could achieve the 64/32 behavior on anycpu by turning those ifdefs into a runtime check.
 
@jkotas what do you think?